src/adapter/src/coord/read_policy.rs

```diff
@@ -281,122 +401,125 @@ impl<S: Append + 'static> crate::coord::Coordinator<S> {
                 policy_changes.push((*id, read_needs.policy()));
             }
             compute
                 .set_read_policy(*compute_instance, policy_changes)
                 .await
                 .unwrap();
         }
+
+        read_holds
     }
-    /// Update the timestamp of the read holds on the indicated collections from the
-    /// indicated time within `read_holds` to `new_time`.
+
+    /// Attempt to update the timestamp of the read holds on the indicated collections from the
+    /// indicated times within `read_holds` to `new_time`.
+    ///
+    /// If we are unable to update a read hold at the provided `time` for a specific id, then we
+    /// leave it unchanged.
     ///
     /// This method relies on a previous call to `acquire_read_holds` with the same
     /// `read_holds` argument or a previous call to `update_read_hold` that returned
     /// `read_holds`, and its behavior will be erratic if called on anything else.
     pub(super) async fn update_read_hold(
         &mut self,
-        mut read_holds: ReadHolds<mz_repr::Timestamp>,
+        read_holds: ReadHolds<mz_repr::Timestamp>,
         new_time: mz_repr::Timestamp,
     ) -> ReadHolds<mz_repr::Timestamp> {
-        let ReadHolds {
-            time: old_time,
-            id_bundle:
-                CollectionIdBundle {
-                    storage_ids,
-                    compute_ids,
-                },
-        } = &read_holds;
+        let mut new_read_holds = ReadHolds::new();
+        let mut storage_policy_changes = Vec::new();
+        let mut compute_policy_changes: BTreeMap<_, Vec<_>> = BTreeMap::new();
 
-        // Update STORAGE read policies.
-        let mut policy_changes = Vec::new();
-        for id in storage_ids.iter() {
-            let collection = self.controller.storage.collection(*id).unwrap();
-            assert!(collection
-                .read_capabilities
-                .frontier()
-                .less_equal(&new_time),
-                "Storage collection {:?} has read frontier {:?} not less-equal new time {:?}; old time: {:?}",
-                id,
-                collection.read_capabilities.frontier(),
-                new_time,
-                old_time,
-            );
-            let read_needs = self.read_capability.get_mut(id).unwrap();
-            read_needs.holds.update_iter(Some((new_time, 1)));
-            read_needs.holds.update_iter(Some((*old_time, -1)));
-            policy_changes.push((*id, read_needs.policy()));
+        for (old_time, id_bundle) in read_holds.holds {
+            if old_time < new_time {
+                new_read_holds
+                    .holds
+                    .entry(new_time)
+                    .or_default()
+                    .extend(&id_bundle);
+                for id in id_bundle.storage_ids {
+                    let collection = self.controller.storage.collection(id).unwrap();
+                    assert!(collection
```

Comment:
I think this is a valid assert. If we think that we currently have some read hold at `old_time`, and the `old_time` is less than `new_time`, then the read frontier should be less than `new_time`.

Thoughts? Is it possible that even though we think we have a read hold, the read frontier still advances? Maybe after a compute/storage reboot this is possible, I'm not sure.

Reply:
After speaking with Frank in person, the new assert should be valid.